### PR TITLE
Check AVX, AVX2, AVX512XEON compiler macros on KokkosCore_config.h

### PR DESCRIPTION
**Pull request by masterleinad**
We discussed that we also want to have the reverse check as the one in #6188, i.e., check that we actually see the expected compiler macro when setting an architecture (for `AVX`, `AVX2`, `AVX512XEON` in this pull request).

### CODE REVIEW

**Comment by crtrott** [comment]
What do other folks think about the place where to check this? We could also do it in Kokkos_Macros.hpp, or have a separate file included in Kokkos_Macros.hpp (i.e. add these checks inside: Kokkos_Architecture_Macro_Verification.hpp).

**Comment by dalg24** [comment]
I am not sure about the direction.  I would like to look into dropping these `KOKKOS_ARCH_AVX[2,512XEON]` macros

**Comment by crtrott** [comment]
I am pretty sure we need a way of testing. Say you build Trilinos with AVX512, but then you build an App on top of it with only AVX2. That would lead to issues we should catch.

**Comment by masterleinad** [comment]
> I am not sure about the direction. I would like to look into dropping these `KOKKOS_ARCH_AVX[2,512XEON]` macros

If you don't have these (or similar macros), you could only detect SIMD support directly through the compiler macros and that is the approach we moved away from in https://github.com/kokkos/kokkos/pull/6188 to avoid ODR violations/compiling with different architecture flags.

**Comment by masterleinad** [comment]
Retest this please.

**Review by dalg24 — COMMENTED** [review]
These seem out of place.
Introduce an `impl/Kokkos_AdvancedVectorExtensionsMacros.hpp` file or whatnot.

**Review by dalg24 — APPROVED** [review]

File: simd/src/impl/Kokkos_AdvancedVectorExtensionsMacros.hpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+#ifndef KOKKOS_ADVANCED_VECTOR_EXTENSIONS_MACROS_HPP
+#define KOKKOS_ADVANCED_VECTOR_EXTENSIONS_MACROS_HPP
+
+#include <Kokkos_Macros.hpp>
+
+#if defined(KOKKOS_ARCH_AVX) && !defined(__AVX__)
+#error "__AVX__ must be defined for KOKKOS_ARCH_AVX"
+#endif
+#cmakedefine KOKKOS_ARCH_AVX2
```
Comment (dalg24) [inline simd/src/impl/Kokkos_AdvancedVectorExtensionsMacros.hpp:25]:
```suggestion

```

**Review by dalg24 — CHANGES_REQUESTED** [review]

File: simd/src/impl/Kokkos_AdvancedVectorExtensionsMacros.hpp (hunk 1 above)
Comment (masterleinad) [reply simd/src/impl/Kokkos_AdvancedVectorExtensionsMacros.hpp]: Fixed.

**Review by ldh4 — COMMENTED** [review]
A few thoughts.
Will there be a separate check for NEON?

I think changing the file name to something like `Kokkos_AVXMacrosCheck.hpp` could be useful as well, since the file doesn't really introduce any macros about AVX.

Also, I think this part could be moved and incorporated into the new file as part of its check:
```
#ifdef KOKKOS_ARCH_AVX2
#include <Kokkos_SIMD_AVX2.hpp>
#endif
#ifdef KOKKOS_ARCH_AVX512XEON
#include <Kokkos_SIMD_AVX512.hpp>
#endif
```

**Comment by masterleinad** [comment]
> Will there be a separate check for NEON?

Yes, once we have this one and #6188 in, I wanted to propose the respective changes for `NEON` that are somewhat backward-incompatible.

File: simd/src/impl/Kokkos_AdvancedVectorExtensionsMacros.hpp (hunk 1 above)
Comment (dalg24) [reply simd/src/impl/Kokkos_AdvancedVectorExtensionsMacros.hpp]: I meant the empty line but anyway I think DongHun's suggestion is better.

**Comment by masterleinad** [comment]
I moved the checks inside `Kokkos_SIMD.hpp`. I like that better than creating a separate file as well.

File: simd/src/Kokkos_SIMD.hpp [hunk 2]
```diff
@@ -21,11 +21,17 @@
...
-#ifdef KOKKOS_ARCH_AVX2
+#include <Kokkos_Macros.hpp>
+
+#if defined(KOKKOS_ARCH_AVX) && !defined(__AVX__)
+#error "__AVX__ must be defined for KOKKOS_ARCH_AVX"
+#endif
+#if defined(KOKKOS_ARCH_AVX2) && !defined(__AVX2__)
+#error "__AVX2__ must be defined for KOKKOS_ARCH_AVX2"
 #include <Kokkos_SIMD_AVX2.hpp>
 #endif
```
Comment (ldh4) [inline simd/src/Kokkos_SIMD.hpp:32]: I agree with moving the check into `Kokkos_SIMD.hpp`, but is the condition to include each header correct?

Shouldn't it be something like this?
```suggestion
#if defined(KOKKOS_ARCH_AVX2)
#if !defined(__AVX2__)
#error "__AVX2__ must be defined for KOKKOS_ARCH_AVX2"
#else
#include <Kokkos_SIMD_AVX2.hpp>
#endif
#endif
```

Comment (masterleinad) [reply simd/src/Kokkos_SIMD.hpp]: Of course, fixed!

**Review by ldh4 — APPROVED** [review]

File: simd/src/Kokkos_SIMD.hpp [hunk 3]
```diff
@@ -21,11 +21,21 @@
 
 #include <Kokkos_SIMD_Scalar.hpp>
 
-#ifdef KOKKOS_ARCH_AVX2
+#include <Kokkos_Macros.hpp>
+
+#if defined(KOKKOS_ARCH_AVX) && !defined(__AVX__)
+#error "__AVX__ must be defined for KOKKOS_ARCH_AVX"
+#endif
```
Comment (dalg24) [inline simd/src/Kokkos_SIMD.hpp:28]: Break a line plz

Comment (masterleinad) [reply simd/src/Kokkos_SIMD.hpp:28]: Here you go!

**Review by dalg24 — APPROVED** [review]

**Comment by masterleinad** [comment]
`CUDA-12.2-NVHPC`, `HIP-ROCm-5.2`, and `HIP-ROCm-5.6-C++20` were timing out. Everything else passed.

File: simd/src/Kokkos_SIMD.hpp [hunk 4]
```diff
@@ -21,11 +21,23 @@
 
 #include <Kokkos_SIMD_Scalar.hpp>
 
-#ifdef KOKKOS_ARCH_AVX2
+#include <Kokkos_Macros.hpp>
```
Comment (dalg24) [inline simd/src/Kokkos_SIMD.hpp:24]: This header include is unnecessary but whatever

**Pull request merged**